// bizcharts/es/extend/scale/scale.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _get2 = _interopRequireDefault(require("@babel/runtime/helpers/get"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _scale = require("@antv/scale");

var _linearTick = _interopRequireDefault(require("./linear-tick"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

(0, _scale.registerTickMethod)('linear-strict-tick-method', _linearTick["default"]);
/**
 * 线性度量
 * @class
 */

var LinearStrict = /*#__PURE__*/function (_Linear) {
  (0, _inherits2["default"])(LinearStrict, _Linear);

  var _super = _createSuper(LinearStrict);

  function LinearStrict(props) {
    var _this;

    (0, _classCallCheck2["default"])(this, LinearStrict);
    _this = _super.call(this, props);
    _this.type = 'linear-strict';
    return _this;
  }

  (0, _createClass2["default"])(LinearStrict, [{
    key: "initCfg",
    value: function initCfg() {
      (0, _get2["default"])((0, _getPrototypeOf2["default"])(LinearStrict.prototype), "initCfg", this).call(this);
      this.tickMethod = 'linear-strict-tick-method';
    }
  }, {
    key: "calculateTicks",
    value: function calculateTicks() {
      var preNice = this.nice; // 强制按照计算出来的ticks，需要设置nice为true，不然会根据最值进行过滤

      this.nice = true;
      var ticks = (0, _get2["default"])((0, _getPrototypeOf2["default"])(LinearStrict.prototype), "calculateTicks", this).call(this);
      this.nice = preNice; // 根据ticks设置最值

      if (ticks.length) {
        this.min = ticks[0];
        this.max = ticks[ticks.length - 1];
      }

      return ticks;
    }
  }]);
  return LinearStrict;
}(_scale.Linear);

var _default = LinearStrict;
exports["default"] = _default;

if (!(0, _scale.getScale)('linear-strict')) {
  (0, _scale.registerScale)('linear-strict', LinearStrict);
}
