// bizcharts/es/boundary/withContainer.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = withContainer;

var _object = _interopRequireDefault(require("object.assign"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _react = _interopRequireWildcard(require("react"));

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) {
    if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0) t[p] = s[p];
  }

  if (s != null && typeof Object.getOwnPropertySymbols === "function") for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
    if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

function withContainer(Component) {
  var name = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'ChartContainer';

  var Cls = /*#__PURE__*/_react["default"].forwardRef(function (props, ref) {
    var container = (0, _react.useRef)();

    var _useState = (0, _react.useState)(false),
        _useState2 = (0, _slicedToArray2["default"])(_useState, 2),
        inited = _useState2[0],
        setInited = _useState2[1];

    var _a = props,
        _a$className = _a.className,
        className = _a$className === void 0 ? "bizcharts" : _a$className,
        containerStyle = _a.containerStyle,
        options = __rest(_a, ["className", "containerStyle"]);

    (0, _react.useEffect)(function () {
      setInited(true);
    }, []); // @ts-ignore

    return /*#__PURE__*/_react["default"].createElement("div", {
      ref: container,
      className: className,
      // @ts-ignore
      style: (0, _object["default"])({
        position: 'relative',
        height: props.height || '100%',
        width: props.width || '100%'
      }, containerStyle)
    }, inited ? /*#__PURE__*/_react["default"].createElement(Component, (0, _object["default"])({
      ref: ref,
      container: container.current
    }, options)) : /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null));
  });

  Cls.displayName = name || Component.name;
  return Cls;
}

;
